xds/internal/testutils/fakeclient/client.go

```diff
@@ -32,57 +31,94 @@ import (
 	closeCh      *testutils.Channel
 	loadStore    *load.Store
 
-	mu        sync.Mutex
-	serviceCb func(xdsclient.ServiceUpdate, error)
-	cdsCb     func(xdsclient.ClusterUpdate, error)
-	edsCb     func(xdsclient.EndpointsUpdate, error)
+	ldsCb func(xdsclient.ListenerUpdate, error)
+	rdsCb func(xdsclient.RouteConfigUpdate, error)
+	cdsCb func(xdsclient.ClusterUpdate, error)
+	edsCb func(xdsclient.EndpointsUpdate, error)
 }
 
-// WatchService registers a LDS/RDS watch.
-func (xdsC *Client) WatchService(target string, callback func(xdsclient.ServiceUpdate, error)) func() {
-	xdsC.mu.Lock()
-	defer xdsC.mu.Unlock()
-
-	xdsC.serviceCb = callback
-	xdsC.suWatchCh.Send(target)
+// WatchListener registers a LDS watch.
+func (xdsC *Client) WatchListener(serviceName string, callback func(xdsclient.ListenerUpdate, error)) func() {
+	xdsC.ldsCb = callback
+	xdsC.ldsWatchCh.Send(serviceName)
 	return func() {
-		xdsC.suCancelCh.Send(nil)
+		xdsC.ldsCancelCh.Send(nil)
 	}
 }
 
-// WaitForWatchService waits for WatchService to be invoked on this client and
+// WaitForWatchListener waits for WatchCluster to be invoked on this client and
 // returns the serviceName being watched.
-func (xdsC *Client) WaitForWatchService(ctx context.Context) (string, error) {
-	val, err := xdsC.suWatchCh.Receive(ctx)
+func (xdsC *Client) WaitForWatchListener(ctx context.Context) (string, error) {
+	val, err := xdsC.ldsWatchCh.Receive(ctx)
 	if err != nil {
 		return "", err
 	}
 	return val.(string), err
 }
 
-// InvokeWatchServiceCallback invokes the registered service watch callback.
-func (xdsC *Client) InvokeWatchServiceCallback(u xdsclient.ServiceUpdate, err error) {
-	xdsC.mu.Lock()
-	defer xdsC.mu.Unlock()
+// InvokeWatchListenerCallback invokes the registered ldsWatch callback.
+//
+// Not thread safe with WatchListener. Only call this after
+// WaitForWatchListener.
+func (xdsC *Client) InvokeWatchListenerCallback(update xdsclient.ListenerUpdate, err error) {
+	xdsC.ldsCb(update, err)
+}
 
-	xdsC.serviceCb(u, err)
+// WaitForCancelListenerWatch waits for a LDS watch to be cancelled  and returns
+// context.DeadlineExceeded otherwise.
+func (xdsC *Client) WaitForCancelListenerWatch(ctx context.Context) error {
+	_, err := xdsC.ldsCancelCh.Receive(ctx)
+	return err
+}
+
+// WatchRoute registers a RDS watch.
+func (xdsC *Client) WatchRoute(routeName string, callback func(xdsclient.RouteConfigUpdate, error)) func() {
+	xdsC.rdsCb = callback
+	xdsC.rdsWatchCh.Send(routeName)
+	return func() {
+		xdsC.rdsCancelCh.Send(nil)
+	}
+}
+
+// WaitForWatchRoute waits for WatchCluster to be invoked on this client and
+// returns the clusterName being watched.
+func (xdsC *Client) WaitForWatchRoute(ctx context.Context) (string, error) {
+	val, err := xdsC.rdsWatchCh.Receive(ctx)
+	if err != nil {
+		return "", err
+	}
+	return val.(string), err
+}
+
+// InvokeWatchRouteCallback invokes the registered rdsWatch callback.
+//
+// Not thread safe with WatchListener. Only call this after
```

Comment:
Should this say `WatchRoute` instead? And similarly for the cluster and endpoints functions.

Reply:
Yes. Done